lib/msf/ui/console/command_dispatcher/modules.rb

```diff
@@ -971,6 +981,196 @@ def cmd_back(*args)
+                next
+              end
+
+              if contents && contents.include?(mod.fullname)
+                print_warning("Module #{mod.fullname} has already been favorited and will not be added to #{favs_file}")
+                next
+              end
+
+              File.open(favs_file, 'a+') { |file| file.puts(mod.fullname) }
+              print_status("Added #{mod.fullname} to #{favs_file}")
+            end
+            return
+          end
+
+          #
+          # Helper method for cmd_favorite that deletes modules from the fav_modules_file, or deletes the file altogether
+          #
+          def favorite_del(modules, delete_all, favs_file, contents)
+            if delete_all
+              print_status("Clearing the contents of #{favs_file}")
+              File.write(favs_file, '')
+              return
+            end
+
+            modules = modules.uniq # prevent modules from being deleted more than once
+            contents = contents.split
+            modules.each do |name|
+              mod = framework.modules.create(name)
+              if (mod == nil)
+                print_error("Invalid module: #{name}")
+                next
+              end
+
+              unless contents.include?(mod.fullname)
+                print_warning("Module #{mod.fullname} cannot be deleted because it is not in #{favs_file}")
+                next
+              end
+
+              # remove module from contents
+              contents.delete(mod.fullname)
+              print_status("Removing #{mod.fullname} from #{favs_file}")
+            end
+
+            # clear the contents of the fav_modules file if removing the module(s) makes it empty
+            if contents.length == 0
+              print_status("Clearing the contents of #{favs_file} because there are no modules left")
+              File.write(favs_file, '')
+              return
+            end
+            File.open(favs_file, 'w') { |file| file.puts(contents.join("\n")) }
+          end
+
+          #
+          # Helper method for cmd_favorite that checks if the fav_modules file exists and is readable / writable
+          #
+          def favorite_check_fav_modules(favs_file)
+            exists = false
+            writable = false
+            readable = false
+            contents = ''
+            return_array = [exists, writable, readable, contents]
+            
+            return return_array unless File.exists?(favs_file)
+
+            exists = true
+
+            if File.writable?(favs_file)
+              writable = true
+            end
+
+            if File.readable?(favs_file)
+              readable = true
+              contents = File.read(favs_file)
+            end
+
+            return_array = [exists, writable, readable, contents]
+            return return_array
+          end
+
+          #
+          # Add modules to or delete modules from the fav_modules file
+          #
+          def cmd_favorite(*args)
```

Comment:
The implementation so far is great, although there are a few things that I think would make it better:

1. Since the file permissions checking is the same between the `favorite_add()` and `favorite_del()` methods and the `contents` variable is never used in this method aside from passing it to the helper methods, it might be better to call `favorite_check_fav_modules()` in `favorite_add()` and `favorite_del()` before altering the favorites file.
2. There are a few places where a user might enter an invalid option or text. In those places, let's tell the user why the entered options/arguments are wrong and then call `cmd_favorite_help()` to display the usage for them.
3. I mistakenly suggested using `parse()` as I thought it supported multiple tokens for a given argument. It seems I was wrong there, so your initial way of checking the arguments is definitely the better route to take.

This is untested, but this implements the improvements I listed above, and it assumes the file permissions checking / reading contents is done via the `favorite_add()` and `favorite_del()` methods

<details>
<summary>code</summary>

```ruby
def cmd_favorite(*args)
  favs_file = Msf::Config.fav_modules_file

  if args.include?('-h') || (args.include?('-c') && args.include?('-d'))
    cmd_favorite_help
    return
  end
	
  if args.empty?
    unless active_module
      print_error('No module has been provided to favorite.')
      cmd_favorite_help
      return
    end

    args = [ active_module.fullname ]
    favorite_add(args, favs_file)
    return
  end

  case args[0]
  when '-c'
    args.delete('-c')
    unless args.empty?
      print_error('Option -c does not support arguments.')
      cmd_favorite_help
      return
    end
		
    favorite_del(args, true, favs_file)
  when '-d'
    args.delete('-d')
    if args.empty?
      unless active_module
        print_error('No module has been provided to delete.')
        cmd_favorite_help
        return
      end

      args = [ active_module.fullname ]
    end
	
    favorite_del(args, false, favs_file)
  else # no valid opts, but there are args
    if args[0].start_with?('-')
      print_error('Invalid option provided')
      cmd_favorite_help
      return
    end

    favorite_add(args, favs_file)
  end
end
```
</details>

Reply:
Thanks for the template, this is very neat!